rts/lra/service-base/src/main/java/io/narayana/lra/APIVersion.java

```diff
@@ -0,0 +1,149 @@
+     *                       {@link LRAConstants#NARAYANA_LRA_API_VERSION} is returned
+     * @return instance of the {@link APIVersion} class based on the parsed String
+     * @throws IllegalArgumentException thrown when version string has a wrong format
+     */
+    public static APIVersion instanceOf(String versionString) {
+        if (versionString == null || versionString.isEmpty()) {
+            return LRAConstants.NARAYANA_LRA_API_VERSION;
+        }
+        Matcher versionMatcher = versionPattern.matcher(versionString);
+        if (!versionMatcher.matches()) {
+            throw new IllegalArgumentException("Cannot parse provided version string " + versionString
+                    + " as it does not match the expected format '" + expectedFormat + "'");
+        }
+        try {
+            int major = Integer.parseInt(versionMatcher.group(1));
+            int minor = Integer.parseInt(versionMatcher.group(2));
+            String preRelease = versionMatcher.group(3);
+            return new APIVersion(major, minor, preRelease);
+        } catch (NumberFormatException nfe) {
+            throw new IllegalArgumentException("The version string " + versionString + " matches the expected format " + expectedFormat
+                    + " but the major.minor cannot be converted to numbers", nfe);
+        }
+    }
+
+    private APIVersion(int major, int minor, String preRelease) {
+        this.major = major;
+        this.minor = minor;
+        this.preRelease = preRelease;
+    }
+
+    /**
+     * The 'major' and 'minor' versions are compared in numeric comparison.
+     * If the 'preRelease' exists then the version is considered equal
+     *
+     * If one version does not contain the 'preRelease' version string
+     * then it's considered bigger,
+     * if both versions contains the 'preRelease' strings then they're
+     * compared with {@link String#compareTo(String)} method.
+     *
+     * @param anotherVersion version to compare against
+     * @return < 0 if this version is lower than otherVersion, > 0 if this version is bigger than otherVersion,
+     *         0 if the versions are equal
+     */
+    @Override
+    public int compareTo(APIVersion anotherVersion) {
+        int result = Integer.compare(major, anotherVersion.major);
+        if (result == 0) {
+            result = Integer.compare(minor, anotherVersion.minor);
+        }
+        if (result == 0 ) {
+            if (preRelease == null && anotherVersion.preRelease == null) {
+                result = 0;
+            // version without preRelease is bigger than the one with the preRelease
+            } else if (preRelease == null && anotherVersion.preRelease != null) {
+                result = 1;
+            } else if (preRelease != null && anotherVersion.preRelease == null) {
+                result = -1;
+            } else {
+                result = preRelease.compareTo(anotherVersion.preRelease);
```

Comment:
It would break if they/we changed the format/naming of non-final versions (for example `RCy` versus `Mx`).
BTW I made a comment previously checking for unofficial versions that was never addressed where I argued that the code for preRelease was unnecessary.

Reply:
https://github.com/jbosstm/narayana/pull/1783#discussion_r577690745

Reply:
https://github.com/jbosstm/narayana/pull/1783#discussion_r576950363